### PR TITLE
[GlobalIsel] Canonicalize G_FCMP

### PR DESCRIPTION
**Pull request by tschuett**
As a side-effect, we start constant folding fcmps.

### CODE REVIEW

File: llvm/lib/CodeGen/GlobalISel/CombinerHelperCompares.cpp [hunk 1]
```diff
@@ -84,3 +115,31 @@ bool CombinerHelper::matchCanonicalizeICmp(const MachineInstr &MI,
...
+  if (auto CLHS = GFConstant::getConstant(LHS, MRI)) {
+    if (auto CRHS = GFConstant::getConstant(RHS, MRI))
+      return constantFoldFCmp(*Cmp, *CLHS, *CRHS, MatchInfo);
+
+    // If we have a constant, make sure it is on the RHS.
+    std::swap(LHS, RHS);
+    Pred = CmpInst::getSwappedPredicate(Pred);
+
+    MatchInfo = [=](MachineIRBuilder &B) {
+      B.buildFCmp(Pred, Dst, LHS, RHS, Cmp->getFlags());
```
Comment (tschuett) [inline llvm/lib/CodeGen/GlobalISel/CombinerHelperCompares.cpp:139]: confidential

**Review by aemerson — COMMENTED** [review]
Github won't let me inline comment on code that wasn't modified, but in `llvm::GIConstant::getConstant()` where you call `getIConstantVRegValWithLookThrough()`: it just occurred to me that we could end up with the width of the APInt being different from the width of the original `Register Const` if the look-through went through some SEXT/ZEXT operations. Do you agree?

File: llvm/lib/CodeGen/GlobalISel/Utils.cpp [hunk 2]
```diff
@@ -2008,3 +2008,43 @@ llvm::GIConstant::getConstant(Register Const, const MachineRegisterInfo &MRI) {
...
+  return Values[0];
+}
+
+std::optional<GFConstant>
+llvm::GFConstant::getConstant(Register Const, const MachineRegisterInfo &MRI) {
+  MachineInstr *Constant = getDefIgnoringCopies(Const, MRI);
+
+  if (GSplatVector *Splat = dyn_cast<GSplatVector>(Constant)) {
+    std::optional<FPValueAndVReg> MayBeConstant =
+        getFConstantVRegValWithLookThrough(Splat->getScalarReg(), MRI);
```
Comment (tschuett) [inline llvm/lib/CodeGen/GlobalISel/Utils.cpp:2024]: There are new look-throughs with APFloat this time.

**Comment by tschuett** [comment]
I remember there was a discussion on this topic. IR is well formed and it should work as expected.

const(i128)  = G_ZEXT G_CONSTANT i64

We have an i64 G_CONSTANT and zext it into the i128 const register.

**Comment by aemerson** [comment]
> I remember there was a discussion on this topic. IR is well formed and it should work as expected.
> 
> const(i128) = G_ZEXT G_CONSTANT i64
> 
> We have an i64 G_CONSTANT and zext it into the i128 const register.

Right but is that really what we want by default? What kind of guarantees does the GIConstant API promise about these cases? Naively a user would assume that if they give a VReg with a s64 type, they expect an APInt of 64bits to be returned if it's a constant, not a G_CONSTANT i8. You could imagine a scenario where the result of this is then used to build another constant of the wrong type resulting in crashes or worse.

**Comment by tschuett** [comment]
Maybe there is a misunderstanding.

I have a vreg s128 and ask for a GIConstant. Then I either get a std::nullopt or a wrapper around an APint with 128 bit. Even if between the vreg and the G_CONSTANT are some copies, exts, and trunc. With lookthrough we interpret the IR and get the correct result because IR is always well-formed.


I have a vreg 10xs64 and ask for a GIConstant. Then I either get std::nullopt or there is a G_BUILD_VECTOR directly on the vreg or behind some copies.

**Comment by tschuett** [comment]
We zext the APInt:
https://github.com/llvm/llvm-project/blob/30d7dcc1db476b828d3b0a7b5c9bcfbe5bd5f216/llvm/lib/CodeGen/GlobalISel/Utils.cpp#L369

**Comment by aemerson** [comment]
> We zext the APInt:
> 
> https://github.com/llvm/llvm-project/blob/30d7dcc1db476b828d3b0a7b5c9bcfbe5bd5f216/llvm/lib/CodeGen/GlobalISel/Utils.cpp#L369

Right, I just saw that and I agree it handles it.

**Review by aemerson — APPROVED** [review]

**Pull request merged**